qiskit_optimization/algorithms/qrao/magic_rounding.py

```diff
@@ -0,0 +1,448 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Magic basis rounding module"""
+from dataclasses import dataclass
+
+from collections import defaultdict
+from typing import Any, Dict, List, Optional, Tuple
+
+import numpy as np
+from qiskit import QuantumCircuit
+from qiskit.algorithms.exceptions import AlgorithmError
+from qiskit.primitives import Sampler
+from qiskit.quantum_info import SparsePauliOp
+
+from qiskit_optimization.algorithms import OptimizationResultStatus, SolutionSample
+from .quantum_random_access_encoding import (
+    _z_to_21p_qrac_basis_circuit,
+    _z_to_31p_qrac_basis_circuit,
+)
+from .rounding_common import RoundingContext, RoundingResult, RoundingScheme
+
+
+@dataclass
+class MagicRoundingResult(RoundingResult):
+    """Result of magic rounding."""
+
+    bases: np.ndarray
+    """The bases used for the magic rounding"""
+    basis_shots: np.ndarray
+    """The number of shots used for each basis"""
+    basis_counts: List[Dict[str, int]]
+    """The basis_counts represents the resulting counts obtained by measuring with the bases
+    corresponding to the number of shots specified in basis_shots."""
+
+
+class MagicRounding(RoundingScheme):
+    """Magic rounding scheme
+
+    This scheme is described in https://arxiv.org/abs/2111.03167v2.
+    """
+
+    _DECODING = {
+        3: (  # Eq. (8)
+            {"0": [0, 0, 0], "1": [1, 1, 1]},  # I mu+ I, I mu- I
+            {"0": [0, 1, 1], "1": [1, 0, 0]},  # X mu+ X, X mu- X
+            {"0": [1, 0, 1], "1": [0, 1, 0]},  # Y mu+ Y, Y mu- Y
+            {"0": [1, 1, 0], "1": [0, 0, 1]},  # Z mu+ Z, Z mu- Z
+        ),
+        2: (  # Sec. VII
+            {"0": [0, 0], "1": [1, 1]},  # I xi+ I, I xi- I
+            {"0": [0, 1], "1": [1, 0]},  # X xi+ X, X xi- X
+        ),
+        1: ({"0": [0], "1": [1]},),
+    }
+
+    # Pauli op string to label index in ops
+    _OP_INDICES = {1: {"Z": 0}, 2: {"X": 0, "Z": 1}, 3: {"X": 0, "Y": 1, "Z": 2}}
+
+    def __init__(
+        self,
+        sampler: Sampler,
+        basis_sampling: str = "uniform",
+        seed: Optional[int] = None,
```

Comment:
```suggestion
        seed: int | None = None,
```
We have been preferring to use the vertical bar syntax nowadays